### DISCUSSION TITLE
Error: self signed certificate in certificate chain

### DISCUSSION THREAD

**Discussion by ShrihariKale1** [community/community#39739] [Code Security]
### Select Topic Area

Bug

### Body

CodeQL Workflow - running with self hosted runner however getting as error for action-  github/codeql-action/init@v2

Error:  - _"Downloading CodeQL tools from https://github.com/github/codeql-action/releases/download/codeql-bundle-20221024/codeql-bundle-win64.tar.gz. This may take a while.
Error: self signed certificate in certificate chain"_

1. Added certificates in trusted certificates authority, however same error occurred. 
2. set env. variable NODE_EXTRA_CA_CERTS=<path to pem file>
      getting error - _Error: unable to get issuer certificate_
3. set env. variable NODE_TLS_REJECT_UNAUTHORIZED=0
      getting error - "_aborted_" 

Is there any other way to resolve this error?

**Comment by ossanna16** [comment]
Hi there @ShrihariKale1 and welcome to our community! Thank you for asking a great question 🙂
While you are waiting for a reply to your question, here are a few things you can do:

To get started, introduce yourself in our [official introduction thread](https://github.com/orgs/community/discussions/23204)
To connect with other new users, check out this Discussion to [Meet Your GitHub Community Team](https://github.com/orgs/community/discussions/28776), further your GitHub knowledge by working through our [GitHub Skill 🆙 exercises](https://github.com/orgs/community/discussions/32121), and participate in our [weekly Fun Friday chats](https://github.com/orgs/community/discussions/30912)

**Comment by tausbn** [comment]
Thank you for your question! 

First of all, are you still experiencing this issue? If yes, then it would be helpful to get a few additional pieces of information so that we can (hopefully) figure out what's going on:

- Is this self-hosted runner used with GitHub.com or with a GitHub Enterprise Server instance?
- Is the server private or public?
- Is there some sort of network proxy involved?
- Is the issue intermittent or consistent?

**Reply by ShrihariKale1** [reply]
Yes, issue is not resolved yet. please find required info -
_Is this self-hosted runner used with GitHub.com or with a GitHub Enterprise Server instance?_
- GitHub Enterprise Server instance
_Is the server private or public?_
- Private
_Is there some sort of network proxy involved?_
- Yes, Proxy involded and github domain is whitelisted. And Github certificate chain is added in certificate store on server.
_Is the issue intermittent or consistent?_
- consistent

**Reply by tausbn** [reply]
Thank you for the clarification. We think what is happening here is that the proxy server is decrypting the packets it receives (e.g. to inspect their contents) and then re-encrypting and signing them with a self-signed certificate. 

You could try running 

```
yes | openssl s_client -connect github.com:443 -servername github.com -showcerts | openssl x509 -noout -text
```
on the self-hosted runner and it should tell you what certificates it's seeing.

It _should_ output something like the following:
```
depth=2 C = US, O = DigiCert Inc, OU = www.digicert.com, CN = DigiCert Global Root CA
verify return:1
depth=1 C = US, O = DigiCert Inc, CN = DigiCert TLS Hybrid ECC SHA384 2020 CA1
verify return:1
depth=0 C = US, ST = California, L = San Francisco, O = "GitHub, Inc.", CN = github.com
verify return:1
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            05:18:9a:54:eb:e8:c7:e9:03:e0:ab:0d:92:55:45:de
        Signature Algorithm: ecdsa-with-SHA384
        Issuer: C = US, O = DigiCert Inc, CN = DigiCert TLS Hybrid ECC SHA384 2020 CA1
        Validity
            Not Before: Mar 15 00:00:00 2022 GMT
            Not After : Mar 15 23:59:59 2023 GMT
        Subject: C = US, ST = California, L = San Francisco, O = "GitHub, Inc.", CN = github.com
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:4a:b0:93:71:85:21:ec:62:3f:cb:74:c0:46:c8:
                    e7:00:dc:27:4a:32:4b:8a:d5:51:83:08:11:23:52:
                    65:c5:9d:64:75:94:10:9f:99:6d:3f:7b:fb:29:3b:
                    58:b8:37:54:78:4b:b7:3d:1c:77:7e:90:dd:bb:67:
                    23:32:5c:80:d1
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                keyid:0A:BC:08:29:17:8C:A5:39:6D:7A:0E:CE:33:C7:2E:B3:ED:FB:C3:7A

            X509v3 Subject Key Identifier: 
                78:AA:72:C6:71:69:68:14:B5:59:B1:9E:8B:6E:2B:40:87:42:3B:1E
            X509v3 Subject Alternative Name: 
                DNS:github.com, DNS:www.github.com
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl3.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl

                Full Name:
                  URI:http://crl4.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl

            X509v3 Certificate Policies: 
                Policy: 2.23.140.1.2.2
                  CPS: http://www.digicert.com/CPS

            Authority Information Access: 
                OCSP - URI:http://ocsp.digicert.com/
                CA Issuers - URI:http://cacerts.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crt

            X509v3 Basic Constraints: 
                CA:FALSE
            CT Precertificate SCTs: 
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : AD:F7:BE:FA:7C:FF:10:C8:8B:9D:3D:9C:1E:3E:18:6A:
                                B4:67:29:5D:CF:B1:0C:24:CA:85:86:34:EB:DC:82:8A
                    Timestamp : Mar 15 17:06:38.865 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                30:45:02:20:04:7D:70:D9:EF:62:46:5E:2A:CF:49:12:
                                5A:5E:5F:0C:ED:60:1D:B2:2E:1B:5C:F1:B0:74:AD:8D:
                                A4:7B:DF:27:02:21:00:8E:23:0B:F0:76:7B:69:E5:44:
                                3F:F0:E4:52:5B:70:21:2E:1C:EF:75:55:AC:6F:49:82:
                                EA:11:AC:82:3B:81:63
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : 35:CF:19:1B:BF:B1:6C:57:BF:0F:AD:4C:6D:42:CB:BB:
                                B6:27:20:26:51:EA:3F:E1:2A:EF:A8:03:C3:3B:D6:4C
                    Timestamp : Mar 15 17:06:38.843 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                30:46:02:21:00:CD:72:4A:AF:06:1B:A9:EC:6A:40:34:
                                C7:F4:58:FB:9E:CA:D2:FC:BB:F3:C1:68:80:0E:6F:33:
                                59:78:D3:29:3A:02:21:00:C8:DF:12:D1:3A:00:9D:9F:
                                4E:36:02:16:1F:C1:35:52:A4:D6:69:13:C0:14:DA:BC:
                                99:46:99:6A:26:43:26:81
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : B3:73:77:07:E1:84:50:F8:63:86:D6:05:A9:DC:11:09:
                                4A:79:2D:B1:67:0C:0B:87:DC:F0:03:0E:79:36:A5:9A
                    Timestamp : Mar 15 17:06:38.892 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                30:45:02:21:00:F5:48:F9:11:DF:D7:8D:0A:63:FC:CC:
                                50:45:08:F3:1B:45:05:40:B9:A0:5A:57:3B:DE:79:ED:
                                DC:22:7B:BE:C9:02:20:2A:87:7C:12:A5:0F:3E:D1:10:
                                77:55:E5:59:3C:6C:B3:01:5B:54:3A:4B:B1:9A:D9:26:
                                C5:BA:85:57:B3:AC:C5
    Signature Algorithm: ecdsa-with-SHA384
         30:65:02:30:28:95:86:a7:ed:ec:bf:ec:b4:19:af:d5:7a:92:
         44:d7:c4:fe:31:bf:02:a2:88:6e:68:c6:4e:87:0d:6e:e4:3b:
         92:dd:42:b3:b9:11:83:e6:f7:6f:70:bd:22:9b:b9:15:02:31:
         00:ce:dc:b4:62:ab:60:31:ad:d2:91:fc:88:5e:ab:cd:ba:7d:
         43:81:d0:5c:c6:57:90:dc:60:f9:ba:e3:5b:19:da:39:51:93:
         0c:dc:0d:1f:bd:38:f3:eb:3f:85:3f:5b:51
write:errno=32
```

**Reply by ShrihariKale1** [reply]
its returns diff results - 

C:\Users\SK>yes | openssl s_client -connect github.com:443 -servername github.com -showcerts | openssl x509 -noout -text
'yes' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\SK>openssl s_client -connect github.com:443 -servername github.com -showcerts | openssl x509 -noout -text
54260000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl\statem\extensions.c:894:
Could not read certificate from <stdin>
Unable to load certificate

C:\Users\SK>openssl s_client -connect github.com:443 -servername github.com -showcerts
CONNECTED(000001A4)
8C000000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl\statem\extensions.c:894:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 85 bytes and written 323 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 2E4DF79AEC357DCA7E63E957D50348B5B2B4CE1F26FE53C9A8B11B4D4ADCF2A3
    Session-ID-ctx:
    Master-Key:
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1674116561
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: yes
---

C:\Users\SK>

**Reply by smowton** [reply]
For clarity, with markdown removed:

```
## C:\Users\SK>openssl s_client -connect github.com:443 -servername github.com -showcerts
CONNECTED(000001A4)
8C000000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl\statem\extensions.c:894:
## no peer certificate available
## No client certificate CA names sent
## SSL handshake has read 85 bytes and written 323 bytes
Verification: OK
## New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
Protocol : TLSv1.2
Cipher : 0000
Session-ID: 2E4DF79AEC357DCA7E63E957D50348B5B2B4CE1F26FE53C9A8B11B4D4ADCF2A3
Session-ID-ctx:
Master-Key:
PSK identity: None
PSK identity hint: None
SRP username: None
Start Time: 1674116561
Timeout : 7200 (sec)
Verify return code: 0 (ok)
Extended master secret: yes
C:\Users\SK>
```

**Reply by smowton** [reply]
This certainly looks like the proxy is interfering with connections to github.com, making them appear insecure to the download tool. Is it possible to instruct the proxy to pass github.com connections through unaltered?

**Reply by smowton** [reply]
Alternatively if this is not acceptable because this https middlebox needs to inspect traffic passing through, can we get it to re-sign traffic using a certificate that the NodeJS instance running the codeql-init action trusts, perhaps supplying a custom certificate using `NODE_EXTRA_CA_CERTS`?

**Reply by ShrihariKale1** [reply]
Already tried by setting an environment variable NODE_EXTRA_CA_CERTS, error for it - _Error: unable to get issuer certificate_
options tried and errors are mentioned in above thread.

**Reply by smowton** [reply]
Can you paste a more complete log for either the NODE_EXTRA_CA_CERTS or NODE_TLS_REJECT_UNAUTHORIZED cases?

**Reply by ShrihariKale1** [reply]
Run github/codeql-action/init@v2
  with:
    debug: true
    languages: csharp
    token: ***
    matrix: {
    "language": "csharp"
  }
    setup-python-dependencies: true
  
Setup CodeQL tools
  Looked for CodeQL bundle in github/codeql-action on https://code.**** but got error HttpError: Not Found.
  Downloading CodeQL tools from https://github.com/github/codeql-action/releases/download/codeql-bundle-20230105/codeql-bundle-win64.tar.gz. This may take a while.
  unable to get issuer certificate
  Waiting 13 seconds before trying again
  unable to get issuer certificate
  Waiting 18 seconds before trying again
  Error: Error: unable to get issuer certificate
  Error: Unable to download and extract CodeQL CLI
  Error: Unable to download and extract CodeQL CLI
      at setupCodeQL (D:\actions-runner\IAM-runner\_work\_actions\github\codeql-action\v2\lib\codeql.js:330:15)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async initCodeQL (D:\actions-runner\IAM-runner\_work\_actions\github\codeql-action\v2\lib\init.js:35:38)
      at async run (D:\actions-runner\IAM-runner\_work\_actions\github\codeql-action\v2\lib\init-action.js:98:34)
      at async runWrapper (D:\actions-runner\IAM-runner\_work\_actions\github\codeql-action\v2\lib\init-action.js:179:9)

**Reply by aibaars** [reply]
Another thing to try would be to mirror the codeql-action on your GitHub Enterprise server using https://github.com/github/codeql-action-sync-tool . This is the best way to use CodeQL in environments where access to the public internet is restricted.

**Reply by smowton** [reply]
I believe this would usually happen if the certificate recorded in the NODE_EXTRA_CA_CERTS file is the server certificate or an intermediate certificate, not the self-signed root certificate that NodeJS was complaining about in the first place with the _self signed certificate in certificate chain_ message.

Are you able to try setting NODE_EXTRA_CA_CERTS to point at the self-signed root certificate that (perhaps indirectly) signs the certificate used by the middlebox that's intercepting and re-encrypting HTTPS connections?
